test/integration/a_download_only_test.go

```diff
@@ -0,0 +1,89 @@
+// +build integration
+
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// a_download_only_test.go filename starts with a, for the purpose that it runs before all parallel tests and downloads the images and caches them.
+package integration
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/go-getter"
+	"github.com/pkg/errors"
+	"k8s.io/minikube/pkg/minikube/constants"
+	pkgutil "k8s.io/minikube/pkg/util"
+	"k8s.io/minikube/test/integration/util"
+)
+
+// Note this test runs before all because filename is alpahbetically first
+// is used to cache images and binaries used by other parallel tests to avoid redownloading.
+// TestDownloadOnly tests the --download-only option
+func TestDownloadOnly(t *testing.T) {
+	p := profileName(t)
+	mk := NewMinikubeRunner(t, p)
+	if !isTestNoneDriver() { // none driver doesnt need to be deleted
+		defer mk.TearDown(t)
+	}
+
+	t.Run("Oldest", func(t *testing.T) {
+		stdout, stderr, err := mk.Start("--download-only", fmt.Sprintf("--kubernetes-version=%s", constants.OldestKubernetesVersion))
+		if err != nil {
+			t.Errorf("%s minikube --download-only failed : %v\nstdout: %s\nstderr: %s", p, err, stdout, stderr)
+		}
+	})
+
+	t.Run("Newest", func(t *testing.T) {
+		stdout, stderr, err := mk.Start("--download-only", fmt.Sprintf("--kubernetes-version=%s", constants.NewestKubernetesVersion))
+		if err != nil {
+			t.Errorf("%s minikube --download-only failed : %v\nstdout: %s\nstderr: %s", p, err, stdout, stderr)
+		}
+		// TODO: add test to check if files are downloaded
+	})
+
+	t.Run("DownloadLatestRelease", func(t *testing.T) {
+		dest := filepath.Join(*testdataDir, fmt.Sprintf("minikube-%s-%s-latest-stable", runtime.GOOS, runtime.GOARCH))
+		err := downloadMinikubeBinary(dest, "latest")
+		if err != nil {
+			t.Errorf("erorr downloading the latest minikube release %v", err)
+		}
+	})
+}
+
+// downloadMinikubeBinary downloads the minikube binary from github used by TestVersionUpgrade
```

Comment:
Please ensure that TestVersionUpgrade can still function independently of TestDownloadOnly.

Reply:
The reason I made this compromise was, because I had issues when running tests in parallel, for some reason downloading file (because it happens async) was making it imnpossible to download the file in a not-parallell test, so I had to either choose to make trestversionupgrade a not parallel test or do this hack !

the other option was to download this file in our commons.sh in jenkins

Reply:
thats is a good idea ! I will add download if dioesnt exist to test version upgrade !